### PR TITLE
dev-python: various packages from GURU

### PR DESCRIPTION
**Pull request by Alessandro-Barbieri**
added with permission from the authors
Signed-off-by: Alessandro Barbieri <lssndrbarbieri@gmail.com>

### CODE REVIEW

**Comment by Alessandro-Barbieri** [comment]
work in progress

**Comment by Alessandro-Barbieri** [comment]
@AndrewAmmerlaan I made a giant PR

**Comment by Alessandro-Barbieri** [comment]
Currently `pkgcheck scan uvicorn` fails with
`SourcingError: version 0.11.3: failed sourcing ebuild: returned val from gen_metadata was 'False'`
and I don't know why

**Comment by oczkers** [comment]
It would be great to have [starlette](https://github.com/encode/starlette) and [databases](https://github.com/encode/databases) also included here. I'll try make ebuilds now but i'm not too great especially with tests, optional packages etc. so maybe first push to guru for review?
Anyway can You please wait with merge until those two pops out?

**Comment by AndrewAmmerlaan** [comment]
> It would be great to have [starlette](https://github.com/encode/starlette) and [databases](https://github.com/encode/databases) also included here. I'll try make ebuilds now but i'm not too great especially with tests, optional packages etc. so maybe first push to guru for review?
> Anyway can You please wait with merge until those two pops out?

Sure, I'll take a look at those ebuilds if you commit them to GURU's dev branch. What I usually do to get tests working, is first check if upstream has documented the test dependencies in 'requirements.*' or 'setup.py (usually you can find it there) if this is the case it is easy. Just add those deps to `DEPEND="test? ()"`. If not, I usually just try to enable the tests first without any deps, and then if I get an `package not found` exception add that package to deps and repeat until all deps are found. If there are some tests that fail inside the emerge environment (e.g. because they require internet access), you can play around with PATH or PYTHONPATH a bit, sometimes that helps. Usually I also run the test suite as a regular user, this way I can easily see if a test is failing because it is inside of emerge, or for other reasons. If a test keeps failing (even when running as a regular user and all deps are installed) I usually use sed/rm or a patch to remove the problematic test from the test suite.

**Comment by juippis** [comment]
The point was to introduce **needed** dependencies for *gitfs* package. Could you write the dep-tree down here, and first **make sure** gitfs actually builds before we introduce bunch of packages to the tree?

**Comment by oczkers** [comment]
@AndrewAmmerlaan 
It's already pushed to guru, gotta go out now but i'll make it better based on Your info in a few hours but got two questions:
> do not add entries to PYTHON_COMPAT that are not supported by all dependencies

Is it stict rule? I mean this approche is going to delay whole process of adapting new python version a lot. Doesn't emerge guard PYTHON_COMPAT by himself good enought?

Tests and docs are required or just nice to have?

**Comment by juippis** [comment]
> @AndrewAmmerlaan
> It's already pushed to guru, gotta go out now but i'll make it better based on Your info in a few hours but got two questions:
> 
> > do not add entries to PYTHON_COMPAT that are not supported by all dependencies
> 
> Is it stict rule? I mean this approche is going to delay whole process of adapting new python version a lot. Doesn't emerge guard PYTHON_COMPAT by himself good enought?

The package is essentially broken without working dependencies...

> 
> Tests and docs are required or just nice to have?

They are mandatory (provided tests work) if you want your package to land in ::gentoo repo.

**Comment by AndrewAmmerlaan** [comment]
> It would be great to have [starlette](https://github.com/encode/starlette) and [databases](https://github.com/encode/databases) also included here. I'll try make ebuilds now but i'm not too great especially with tests, optional packages etc. so maybe first push to guru for review?
> Anyway can You please wait with merge until those two pops out?

Sure, I'll take a look at those ebuilds if you commit them to GURU's dev branch. What I usually do to get tests working, is first check if upstream has documented the test dependencies in 'requirements.*' or 'setup.py (usually you can find it there) if this is the case it is easy. Just add those deps to `DEPEND="test? ()"`. If not, I usually just try to enable the tests first without any deps, and then if I get an `package not found` exception add that package to deps and repeat until all deps are found. If there are some tests that fail inside the emerge environment (e.g. because they require internet access), you can play around with PATH or PYTHONPATH a bit, sometimes that helps. Usually I also run the test suite as a regular user, this way I can easily see if a test is failing because it is inside of emerge, or for other reasons. If a test keeps failing (even when running as a regular user and all deps are installed) I usually use sed/rm or a patch to remove the problematic test from the test suite.

> It would be great to have [starlette](https://github.com/encode/starlette) and [databases](https://github.com/encode/databases) also included here. I'll try make ebuilds now but i'm not too great especially with tests, optional packages etc. so maybe first push to guru for review?
> Anyway can You please wait with merge until those two pops out?
> 
> EDIT:
> I've just pushed to guru, it's working but probably need few changes and quality of ebuild is low, sorry yum
> 495be87a14e8533f2d77c253cafb0e2b68a23a39 - databases with deps
> 67adc1d0d2dfbd6a7d2d481da244e664fb5113cc - starlette with deps

@oczkers 
I have just taken a look at one of the packages you added:
Some remarks (see also previous comment):
- Please run `repoman -dx full` on every ebuild before committing it. Among other things it will tell you if you have entries in PYTHON_COMPAT which are not supported by all dependencies.
- Also before committing,  please try `FEATURES="test" USE="doc" emerge -1 .....`. This will actually run the tests and compile the docs (if distutils_enable_sphinx is specified, which it should if there are docs in the source code). 
- The ebuild that I just tested, failed to test because the test were not in the pypi tarball. Make sure that the you are using a tarball that contains the tests and docs. (see also above point, emerging with FEATURES="test" will fail if the tests are not present in the tarball)

I made some changes to one of your ebuilds, see git log. (One that one the test suite seems to be broken, so maybe we have to RESTRICT="test" there, or maybe it would work with an older version of pytest). Could you use my commit in GURU, and the comments here, as a starting point to fix the other ones yourself? ( I have some other things that need doing at the moment, but I can take a look later today or tomorrow)

**Comment by AndrewAmmerlaan** [comment]
> @AndrewAmmerlaan
> It's already pushed to guru, gotta go out now but i'll make it better based on Your info in a few hours but got two questions:
> 
> > do not add entries to PYTHON_COMPAT that are not supported by all dependencies
> 
> Is it stict rule? I mean this approche is going to delay whole process of adapting new python version a lot. Doesn't emerge guard PYTHON_COMPAT by himself good enought?

`python3_8 ${package}` will fail if not all the dependencies are also installed with support for python3_8. Emerge will tell you about this yes, but automated dependency checkers like repoman will also complain. And having all these broken dependencies spamming repoman makes it more or less unusable. Please try to make `repoman -dx full` happy (or at least not failing) before committing.  

> Tests and docs are required or just nice to have?

As @juippis already said, they are required in ::gentoo. GURU is less strict, but it is still a good idea to get tests working, because it will make maintaining the package easier in the long run.

**Comment by Alessandro-Barbieri** [comment]
```
sys-fs/gitfs
	dev-python/atomiclong
	dev-python/raven
		>=dev-python/exam-0.5.2
		~dev-python/pytest-pythonpath-0.7.2
		>=dev-python/sanic-0.7.0
			dev-python/httpcore
			dev-python/pytest-benchmark
			dev-python/pytest-sugar
			dev-python/pytest-sanic
				dev-python/sanic
			dev-python/aiofiles
			dev-python/httptools
			dev-python/websockets
				dev-python/sphinxcontrib-trio
				dev-python/sphinx-autodoc-typehints
					dev-python/sphobjinv
						dev-python/timeout-decorator
						dev-python/stdio-mgr
						dev-python/pytest-subtests
			dev-python/httpx
				dev-python/mkautodoc
				dev-python/hstspreload
				dev-python/trio
					dev-python/outcome
				dev-python/sniffio
					dev-python/curio
				dev-python/uvicorn
					dev-python/httptools
					dev-python/mkdocs-material
						dev-python/pymdown-extensions
					dev-python/black
						dev-python/dataclasses
		dev-python/webpy
			dev-python/dbutils
		dev-python/zconfig
			dev-python/sphinxcontrib-programoutput
```

File: dev-python/uvicorn/uvicorn-0.11.3.ebuild [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	https://pypi.org/project/uvicorn
+"
+SRC_URI="https://github.com/encode/${PN}/archive/${PV}.tar.gz -> ${P}.tar.gz"
+
+LICENSE="BSD"
+KEYWORDS="~amd64 ~x86"
+SLOT="0"
+
+IUSE="doc"
+REQUIRED_USE="x86? ( !doc )"
```
Comment (Alessandro-Barbieri) [inline dev-python/uvicorn/uvicorn-0.11.3.ebuild]: @AndrewAmmerlaan
```[13:50] (@grknight) Baliano: REQUIRED_USE="x86? ( !doc )" is not allowed.  if an arch cannot handle something guarded by a USE, then that USE must be added to the arch's package.use.mask```
BTW why can't build doc on x86?

Comment (juippis) [reply dev-python/uvicorn/uvicorn-0.11.3.ebuild]: I'll look into keywording it for you.

Comment (juippis) [reply dev-python/uvicorn/uvicorn-0.11.3.ebuild]: Done, please rebase your branch against freshly synced ::gentoo repo master.

**Comment by Alessandro-Barbieri** [comment]
a deep dependency doesn't have a license https://github.com/encode/httpcore/issues/47

File: dev-python/uvicorn/uvicorn-0.11.3.ebuild (hunk 1 above)
Comment (AndrewAmmerlaan) [reply dev-python/uvicorn/uvicorn-0.11.3.ebuild]: > Done, please rebase your branch against freshly synced ::gentoo repo master.

Thank you :)

I added some the x86 keyword to some other packages in GURU that depended on mkdocs. And I got rid of the ugly workaround preventing USE="doc" on relevant packages.

**Comment by AndrewAmmerlaan** [comment]
> a dependency doesn't have a license [encode/httpcore#47](https://github.com/encode/httpcore/issues/47)

It is listed as BSD licensed on pypi: https://pypi.org/project/httpcore/

image

**Pull request closed without merging**